Models/ProtoModel.cpp

```diff
@@ -1,105 +1,129 @@
 #include "ProtoModel.h"
 
-#include <QDebug>
+using namespace google::protobuf;
+using CppType = FieldDescriptor::CppType;
 
-ResourceModel::ResourceModel(google::protobuf::Message *protobuf, QObject *parent)
-    : QAbstractItemModel(parent), protobuf(protobuf) {}
+ProtoModel::ProtoModel(Message *protobuf, QObject *parent)
+    : QAbstractItemModel(parent), dirty(false), protobuf(protobuf) {
+  protobufBackup = protobuf->New();
+  protobufBackup->CopyFrom(*protobuf);
+}
+
+void ProtoModel::ReplaceBuffer(Message *buffer) {
+  SetDirty(true);
+  protobuf->CopyFrom(*buffer);
+  emit dataChanged(index(0), index(rowCount()));
+}
+
+void ProtoModel::RestoreBuffer() {
+  std::swap(protobuf, protobufBackup);
+  protobufBackup->CopyFrom(*protobuf);

```

Comment:
@JoshDreamland  says should be protobuf->CopyFrom(*protobufBackup) with no swap

Reply:
The important part of what I said is that both will work, but if this way works and the way you just stated does not, something is probably holding a reference longer than it should be, and that could lead to inexplicable crashes down the road. I would investigate this thoroughly.

Reply:
Yeah that was the very first thing I tried changing about his code, commenting the swap statement and doing ```protobuf->CopyFrom(*protobufBackup)``` but it did not work. I just tried it again, and it still does not work. Only this version we see here on GitHub will show up unchanged on reopen.

Reply:
@JoshDreamland fear not because I came up with a correct fix for it: https://github.com/enigma-dev/RadialGM/commit/b9cae6ef919e27d1850d255c6cf3bf7d6052c53b

The problem was actually just our ImmediateMapper/QDataWidgetMapper forcing the old data back into the widgets because we did not clear the mappings before restoring the buffer (since they are no longer needed) and were not firing a model changed event.